qa/rpc-tests/graphene_stage2.py

```diff
@@ -0,0 +1,95 @@
+                      'graphene_additional_tx_size', 
+                      'graphene_block_size', 
+                      'iblt', 
+                      'inbound_percent', 
+                      'outbound_percent', 
+                      'rank', 
+                      'rerequested', 
+                      'response_time', 
+                      'summary', 
+                      'validation_time'}
+    def __init__(self, test_assertion='success'):
+        self.rep = False
+        BitcoinTestFramework.__init__(self)
+
+        if test_assertion == 'success':
+            self.test_assertion = self.assert_success
+        else:
+            self.test_assertion = self.assert_failure
+
+    def setup_chain(self):
+        print ("Initializing test directory " + self.options.tmpdir)
+        initialize_chain_clean(self.options.tmpdir, 2)
+
+    def setup_network(self, split=False):
+        node_opts = [
+            "-rpcservertimeout=0",
+            "-debug=graphene",
+            "-use-grapheneblocks=1",
+            "-use-thinblocks=0",
+            "-net.grapheneIbltSizeOverride=1",
+            "-net.grapheneBloomFprOverride=0.0",
+            "-net.randomlyDontInv=100",
+            "-excessiveblocksize=6000000",
+            "-blockprioritysize=6000000",
+            "-blockmaxsize=6000000"]
+
+        self.nodes = [
+            start_node(0, self.options.tmpdir, node_opts),
+            start_node(1, self.options.tmpdir, node_opts)
+        ]
+
+        self.is_network_split = False
+        interconnect_nodes(self.nodes)
+        self.sync_all()
+
+    def extract_stats_fields(self, node):
+        gni = node.getnetworkinfo()
+        assert "grapheneblockstats" in gni
+        tbs = gni["grapheneblockstats"]
+        assert "enabled" in tbs and tbs["enabled"]
+        assert set(tbs) == self.expected_stats
+
+        return tbs
+
+    def assert_success(self):
+        # Nodes 1 should have received one block from node 0 and have experienced on decode failure.
+        assert '1 inbound and 0 outbound graphene blocks' in self.extract_stats_fields(self.nodes[1])['summary']
+        assert 'bandwidth with 1 local decode failure' in self.extract_stats_fields(self.nodes[1])['summary']
+
+        # Node 1 should have experienced only one decode failure
+        assert 'bandwidth with 2 local decode failure' not in self.extract_stats_fields(self.nodes[1])['summary']
+
+    def run_test(self):
+        # Generate blocks so we can send a few transactions.  We need some transactions in a block
+        # before a graphene block can be sent and created, otherwise we'll just end up sending a regular
+        # block.
+        self.nodes[0].generate(105)
+        self.sync_blocks()
+
+        logging.info("Send 5 transactions from node0 (to its own address)")
+        addr = self.nodes[0].getnewaddress()
+        for i in range(5):
+            self.nodes[0].sendtoaddress(addr, Decimal("10"))
+
+        self.nodes[0].generate(1)
+
```

Comment:
I wonder if there could be a --extensive type test like we have for parallel.py and excessive.py.  Seems to me that the above will test just the most general case but what about where block sizes change or the number of txns missing changes etc...seems that some sort of randomized longer test would be helpful.

Reply:
I'm not sure that I understand what you mean by block sizes or missing txs changing. Do you mean just that they could be a lot larger? Theoretically this should not influence the correctness of the code (although it might increase the failure rate or present unexpected slowdowns). In general I like the idea of more comprehensive tests or stress tests (I don't actually favor randomized tests if they can be avoided). At any rate, I'd like to advocate for introducing this sort of stress test in a separate PR if that's ok with you.

Reply:
Yes I suppose that's alright. It just seemed to me for the amount of refactoring there isn't much additional testing but perhaps not much really needed here.